src/librustc_resolve/lib.rs

```diff
@@ -2283,24 +2283,25 @@ impl<'a> Resolver<'a> {
                 PatKind::Ident(bmode, ref ident, ref opt_pat) => {
                     // First try to resolve the identifier as some existing
                     // entity, then fall back to a fresh binding.
-                    let resolution = if let Ok(resolution) = self.resolve_path(pat.id,
-                                &Path::from_ident(ident.span, ident.node), 0, ValueNS) {
+                    let local_def = self.resolve_identifier(ident.node, ValueNS, true);
```

Comment:
@petrochenkov 
I agree with that consensus, but haven't gotten around to implementing it yet. Once that's implemented and the inner `x` doesn't resolve to the outer `x` as you described, I agree that it would probably be better to use `resolve_path` or have `resolve_identifier` do the adjustment.